File: python/webblinka/drivers/rtc.py
"""Real-time clocks: the RV-1805, and a base for the rest of the family.

Every I2C RTC worth supporting is the same shape -- a handful of BCD registers
holding the calendar, plus part-specific status bits -- so `RtcDriver` holds
everything that does not vary and a subclass supplies four things: where the
time registers are, how to read and write them, what the part's status bits
mean, and how finely it reports fractions of a second. DS3231, DS1307 and
PCF8563 differ from the RV-1805 in exactly those four ways.

**The panel's job is drift, not the time.** Reading a clock once tells you
nothing about it: a chip that is thirty seconds out and one that gains thirty
seconds a day look identical in a single sample. So the driver keeps its first
reading and reports the rate of change against the host clock in ppm -- along
with the resolution that measurement currently supports, because over a short
session the answer is mostly noise and saying so is more useful than a
confident wrong number.

**Time is UTC here, by decision.** These parts store a bare calendar with no
zone, so somebody has to pick. Reading and writing as UTC means the value is
unambiguous and the arithmetic against the host clock is meaningful; a panel
that showed local time would silently make the drift figure wrong twice a year.
"""

from __future__ import annotations

import calendar
import time
from typing import Any

from .base import Driver, register


def bcd_to_int(value: int) -> int:
    return (value >> 4) * 10 + (value & 0x0F)


def int_to_bcd(value: int) -> int:
    return ((value // 10) << 4) | (value % 10)


class RtcDriver(Driver):
    """Base for I2C real-time clocks."""

    #: Smallest fraction of a second the part reports. Sets how quickly a drift
    #: measurement becomes meaningful: a clock reporting whole seconds needs a
    #: hundred times longer to resolve the same ppm as one reporting hundredths.
    RESOLUTION_S = 1.0
    LABEL = "RTC"

    def __init__(self, bus, address: int) -> None:
        super().__init__(bus, address)
        self._device = None
        self._identity: str | None = None
        self._first: tuple[float, float] | None = None

    # -- to be supplied by the part ----------------------------------------

    def read_clock(self) -> tuple[time.struct_time, float]:
        """The calendar, and the fraction of a second past it."""
        raise NotImplementedError

    def write_clock(self, unix: float) -> None:
        """Set the clock to a unix timestamp.

        Takes the float rather than a struct_time so a part that can store
        fractions of a second is able to. Truncating to whole seconds here
        would leave every clock set by this panel systematically slow by up to
        a second -- which, on a part specified to two parts per million, is
        five centuries' worth of its own error budget.
        """
        raise NotImplementedError

    def flags(self) -> list[dict[str, Any]]:
        """Part-specific status, as rows the panel shows verbatim."""
        return []

    def identify(self) -> str | None:
        """Whatever the part can be asked about itself, if anything."""
        return None

    # -- shared ------------------------------------------------------------

    def start(self) -> dict[str, Any]:
        from adafruit_bus_device.i2c_device import I2CDevice

        self._device = I2CDevice(self.bus, self.address)
        self._first = None
        # Identify before anything reads the calendar. The wrong part at this
        # address still has registers, and they still decode -- into a month of
        # 68 and an exception about the calendar, which tells you nothing about
        # what actually went wrong. Asking first turns that into a sentence
        # naming the part number it found, at the moment you opened the panel.
        self._identity = self.identify()
        return {
            "address": self.address,
            "label": self.LABEL,
            "resolutionS": self.RESOLUTION_S,
            "identity": self._identity,
        }

    def stop(self) -> None:
        self._device = None
        self._first = None

    def command(self, name: str, args: list[Any]) -> Any:
        if name == "set_from_unix":
            # The caller passes the host's idea of now, taken as close to the
            # call as it can manage; the transfer itself is the remaining error.
            self.write_clock(float(args[0]))
            self._first = None  # the old drift baseline describes a different clock
            return self.poll()
        if name == "reset_drift":
            self._first = None
            return self.poll()
        return super().command(name, args)

    def poll(self) -> dict[str, Any]:
        # Bracket the read with host timestamps. The transfer is dozens of HID
        # round-trips, so the device's time corresponds to *somewhere* in that
        # window -- the midpoint is the best estimate and the half-width is the
        # honest uncertainty, which is the floor on any offset figure.
        before = time.time()
        when, fraction = self.read_clock()
        after = time.time()

        host = (before + after) / 2
        uncertainty = max((after - before) / 2, 0.0)
        device = calendar.timegm(when) + fraction
        offset = device - host

        if self._first is None:
            self._first = (host, offset)
        base_host, base_offset = self._first
        elapsed = host - base_host

        drift_ppm = None
        resolution_ppm = None
        if elapsed > 0:
            drift_ppm = ((offset - base_offset) / elapsed) * 1e6
            # What one tick of the part's own resolution would look like over
            # the elapsed time. Below this, the number is quantisation.
            resolution_ppm = (self.RESOLUTION_S / elapsed) * 1e6

        return {
            "label": self.LABEL,
            "address": self.address,
            "iso": _iso(when, fraction),
            "deviceUnix": device,
            "hostUnix": host,
            "offsetS": offset,
            "uncertaintyS": uncertainty,
            "elapsedS": elapsed,
            "driftPpm": drift_ppm,
            "resolutionPpm": resolution_ppm,
            "resolutionS": self.RESOLUTION_S,
            # Read once at start, not per poll: it cannot change, and every
            # register read is HID round-trips.
            "identity": self._identity,
            "flags": self.flags(),
        }

    # -- register helpers --------------------------------------------------

    def _read(self, register: int, length: int) -> bytes:
        buffer = bytearray(length)
        with self._require() as i2c:
            i2c.write_then_readinto(bytes([register]), buffer)
        return bytes(buffer)

    def _write(self, register: int, data: bytes) -> None:
        with self._require() as i2c:
            i2c.write(bytes([register]) + data)

    def _require(self):
        if self._device is None:
            raise RuntimeError(f"{self.LABEL} not started")
        return self._device


def _iso(when: time.struct_time, fraction: float) -> str:
    return "{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:06.3f}Z".format(
        when.tm_year, when.tm_mon, when.tm_mday, when.tm_hour, when.tm_min,
        when.tm_sec + fraction,
    )


# --------------------------------------------------------------- RV-1805


# Register map from the SparkFun RV-1805 library and the RV-1805/AB1805
# datasheet. Time is eight consecutive BCD registers starting at hundredths.
RV1805_HUNDREDTHS = 0x00
RV1805_STATUS = 0x0F
RV1805_CTRL1 = 0x10
RV1805_OSC_STATUS = 0x1D
RV1805_ID0 = 0x28

#: Register 0x28 always reads 0x18 on this part. Worth checking, because 0x69
#: is a busy address -- an MPU-6050 sits there too -- so "something answered"
#: is not the same as "the RTC is there".
RV1805_PART_NUMBER = 0x18


@register("rv1805")
class Rv1805(RtcDriver):
    """Micro Crystal RV-1805 / Abracon AB1805."""

    LABEL = "RV-1805"
    #: This part counts hundredths of a second, which is why it can resolve a
    #: drift rate a hundred times sooner than a whole-seconds clock.
    RESOLUTION_S = 0.01

    def identify(self) -> str | None:
        part = self._read(RV1805_ID0, 1)[0]
        if part != RV1805_PART_NUMBER:
            raise RuntimeError(
                f"whatever is at {self.address:#04x} reports part number "
                f"{part:#04x}, not the {RV1805_PART_NUMBER:#04x} an RV-1805 does"
            )
        return f"part number {part:#04x}"

    def read_clock(self) -> tuple[time.struct_time, float]:
        raw = self._read(RV1805_HUNDREDTHS, 8)
        hundredths, seconds, minutes, hours, date, month, year, weekday = raw
        return (
            time.struct_time(
                (
                    2000 + bcd_to_int(year),
                    bcd_to_int(month),
                    bcd_to_int(date),
                    # Bit 6 of Control1 selects 12-hour mode. The panel and this
                    # driver both work in 24-hour, and start() leaves the part
                    # in it, so the hours register is read straight.
                    bcd_to_int(hours & 0x3F),
                    bcd_to_int(minutes),
                    bcd_to_int(seconds),
                    bcd_to_int(weekday & 0x07),
                    0,
                    -1,
                )
            ),
            bcd_to_int(hundredths) / 100,
        )

    def write_clock(self, unix: float) -> None:
        # Round to the nearest hundredth, then let the carry ripple: rounding
        # 12:00:59.999 up must produce 12:01:00.00, not a sixtieth second.
        hundredths = round(unix * 100)
        when = time.gmtime(hundredths // 100)
        self._write(
            RV1805_HUNDREDTHS,
            bytes(
                [
                    int_to_bcd(hundredths % 100),
                    int_to_bcd(when.tm_sec),
                    int_to_bcd(when.tm_min),
                    int_to_bcd(when.tm_hour),
                    int_to_bcd(when.tm_mday),
                    int_to_bcd(when.tm_mon),
                    int_to_bcd(when.tm_year % 100),
                    # struct_time counts Monday as 0; the part counts Sunday.
                    int_to_bcd((when.tm_wday + 1) % 7),
                ]
            ),
        )

    def flags(self) -> list[dict[str, Any]]:
        status = self._read(RV1805_STATUS, 1)[0]
        control = self._read(RV1805_CTRL1, 1)[0]
        oscillator = self._read(RV1805_OSC_STATUS, 1)[0]

        # OMODE says which oscillator is actually running. Falling back to the
        # RC costs about three orders of magnitude of accuracy, and nothing
        # about the time it reports looks any different -- so it is the first
        # thing to check when a clock is drifting absurdly.
        on_rc = bool(oscillator & (1 << 4))
        failed = bool(oscillator & (1 << 1))

        return [
            {
                "label": "Oscillator",
                "value": "RC (internal)" if on_rc else "XT (32.768 kHz crystal)",
                "tone": "warn" if on_rc else "ok",
            },
            {
                "label": "Oscillator fault",
                "value": "yes — the time is not trustworthy" if failed else "no",
                "tone": "error" if failed else "ok",
            },
            {
                "label": "Power",
                "value": "running from backup" if status & (1 << 6) else "main supply",
                "tone": "warn" if status & (1 << 6) else "ok",
            },
            {
                "label": "Backup battery",
                "value": "low" if status & (1 << 4) else "ok",
                "tone": "warn" if status & (1 << 4) else "ok",
            },
            {
                "label": "Counting",
                "value": "stopped" if control & (1 << 7) else "running",
                "tone": "error" if control & (1 << 7) else "ok",
            },
        ]

File: src/ui/panels/rtc.ts
import type { DevicePanel, DeviceSession } from "../../devices/panel.ts";
import { el } from "../dom.ts";
import { panel, statusPill } from "../panel.ts";

export interface RtcFlag {
  label: string;
  value: string;
  tone: "ok" | "warn" | "error";
  /** A driver command that resolves this flag, when one exists. */
  action?: string | null;
}

export interface RtcState {
  label: string;
  address: number;
  iso: string;
  deviceUnix: number;
  hostUnix: number;
  offsetS: number;
  uncertaintyS: number;
  elapsedS: number;
  driftPpm: number | null;
  resolutionPpm: number | null;
  resolutionS: number;
  identity: string | null;
  flags: RtcFlag[];
}

const POLL_INTERVAL_MS = 1000;

/**
 * Any I2C real-time clock. Nothing here knows which part it is talking to --
 * the driver supplies the time, the status rows and the resolution, so
 * DS3231, DS1307 and PCF8563 need a driver each and no UI at all.
 *
 * It leads with the offset from the host clock rather than the time, because
 * the time is the part of an RTC you can least easily be wrong about: a clock
 * showing the wrong century is obvious, and one losing a second a day looks
 * perfect. The drift rate is the number worth having, so the panel measures it
 * across the session and states the precision that measurement currently
 * supports — over a short window the answer is quantisation, and reporting
 * "±420 ppm so far" is worth more than a confident 3.
 */
export class RtcPanel implements DevicePanel {
  readonly root: HTMLElement;
  readonly #session: DeviceSession;
  readonly #status = statusPill("Reading…", "busy");
  readonly #clock = el("div", { class: "rtc-clock", text: "—" });
  readonly #offset = el("p", { class: "rtc-offset" });
  readonly #drift = el("p", { class: "hint" });
  readonly #flags = el("dl", { class: "facts" });
  readonly #message = el("p", { class: "hint" });
  #timer: number | null = null;
  #polling = false;

  constructor(session: DeviceSession) {
    this.#session = session;
    const p = panel("Clock");
    this.root = p.root;

    const sync = el("button", {
      class: "primary",
      text: "Set from this computer",
      title: "Writes the browser's current UTC into the clock.",
    });
    sync.addEventListener("click", () => void this.#sync(sync));

    const restart = el("button", {
      text: "Restart drift",
      title: "Forget the baseline and start measuring the rate again from now.",
    });
    restart.addEventListener("click", () => void this.#restart(restart));

    p.actions.append(this.#status.node, restart, sync);
    p.body.append(
      this.#clock,
      this.#offset,
      this.#drift,
      this.#flags,
      this.#message,
      el("p", {
        class: "hint",
        text:
          "The clock is read and written as UTC. These parts store a bare " +
          "calendar with no time zone, so something has to choose, and UTC is " +
          "the choice that keeps the drift figure meaningful across a change " +
          "of season.",
      }),
    );
  }

  show(): void {
    if (this.#timer !== null) return;
    this.#timer = setInterval(() => void this.#poll(), POLL_INTERVAL_MS) as unknown as number;
    void this.#poll();
  }

  hide(): void {
    if (this.#timer !== null) clearInterval(this.#timer);
    this.#timer = null;
  }

  async #sync(button: HTMLButtonElement): Promise<void> {
    button.disabled = true;
    try {
      // Sampled here rather than in Python so the value is the host's own
      // clock at the moment of asking; the transfer that follows is the error.
      const state = await this.#session.command<RtcState>("set_from_unix", Date.now() / 1000);
      this.#render(state);
      this.#message.textContent =
        "Set. Anything under a few tens of milliseconds out is the write itself, " +
        "not the clock.";
    } catch (err) {
      this.#fail(err);
    } finally {
      button.disabled = false;
    }
  }

  async #restart(button: HTMLButtonElement): Promise<void> {
    button.disabled = true;
    try {
      this.#render(await this.#session.command<RtcState>("reset_drift"));
      this.#message.textContent = "";
    } catch (err) {
      this.#fail(err);
    } finally {
      button.disabled = false;
    }
  }

  async #poll(): Promise<void> {
    if (this.#polling) return;
    this.#polling = true;
    try {
      this.#render(await this.#session.poll<RtcState>());
    } catch (err) {
      this.#fail(err);
    } finally {
      this.#polling = false;
    }
  }

  #render(state: RtcState): void {
    this.#clock.textContent = state.iso.replace("T", "  ").replace("Z", "");
    this.#status.set(state.identity ?? state.label, "ok");

    const ms = state.offsetS * 1000;
    const uncertainty = state.uncertaintyS * 1000;
    this.#offset.textContent =
      `${ms >= 0 ? "+" : "−"}${Math.abs(ms).toFixed(0)} ms against this computer ` +
      `(±${uncertainty.toFixed(0)} ms, the time the read itself takes)`;
    this.#offset.dataset.tone = Math.abs(ms) > 2000 ? "warn" : "ok";

    // The reasoning about what the drift figure is worth hangs off the offset
    // line rather than taking a paragraph of its own -- it is the same thought,
    // and it is long.
    const drift = describeDrift(state);
    this.#offset.title = drift.detail;
    this.#drift.textContent = drift.summary;
    this.#drift.hidden = drift.summary === "";
    this.#drift.title = drift.detail;
    this.#renderFlags(state.flags);
  }

  #renderFlags(flags: RtcFlag[]): void {
    this.#flags.replaceChildren();
    for (const flag of flags) {
      const value = el("dd", {}, [el("span", { text: flag.value })]);
      value.dataset.tone = flag.tone;
      if (flag.action) {
        // A flag you can do something about should say so where it is raised,
        // not leave you to find the button that clears it.
        const fix = el("button", { class: "inline", text: "Clear" });
        fix.addEventListener("click", () => void this.#act(flag.action!, fix));
        value.append(" ", fix);
      }
      this.#flags.append(el("dt", { text: flag.label }), value);
    }
  }

  async #act(command: string, button: HTMLButtonElement): Promise<void> {
    button.disabled = true;
    try {
      this.#render(await this.#session.command<RtcState>(command));
      this.#message.textContent = "";
    } catch (err) {
      this.#fail(err);
    } finally {
      button.disabled = false;
    }
  }

  #fail(err: unknown): void {
    this.#status.set("Failed", "error");
    this.#message.textContent = err instanceof Error ? err.message : String(err);
  }
}

/**
 * Say what the drift measurement is worth, not just what it is.
 *
 * A rate computed over ten seconds from a clock that counts hundredths can be
 * out by a thousand ppm purely from quantisation, so quoting it to one decimal
 * would be a lie told precisely. The summary is what shows; the detail is the
 * tooltip, because the reasoning is worth having but not worth a paragraph on
 * screen at all times.
 */
function describeDrift(state: RtcState): { summary: string; detail: string } {
  if (state.driftPpm === null || state.resolutionPpm === null || state.elapsedS < 2) {
    return {
      summary: "",
      detail: "A drift rate needs two readings some time apart. Leave it running.",
    };
  }

  const elapsed = formatDuration(state.elapsedS);
  const resolution = state.resolutionPpm;
  if (Math.abs(state.driftPpm) < resolution) {
    return {
      summary: "",
      detail:
        `No drift resolvable yet: ${elapsed} of observation resolves about ` +
        `±${resolution.toFixed(0)} ppm, and the measurement is inside that. ` +
        `Leave it running — the bound improves in proportion to the time.`,
    };
  }

  const perDay = (state.driftPpm * 86400) / 1e6;
  const digits = resolution < 1 ? 2 : 0;
  return {
    summary:
      `Drifting ${state.driftPpm > 0 ? "fast" : "slow"} by ` +
      `${Math.abs(state.driftPpm).toFixed(digits)} ppm — ` +
      `about ${Math.abs(perDay).toFixed(1)} s/day.`,
    detail:
      `${Math.abs(state.driftPpm).toFixed(digits)} ±${resolution.toFixed(digits)} ppm ` +
      `measured over ${elapsed}, which works out at ` +
      `${Math.abs((perDay * 365) / 60).toFixed(0)} min/year. The bound is what ` +
      `one tick of this part's resolution looks like over that long, and it ` +
      `improves in proportion to the time.`,
  };
}

function formatDuration(seconds: number): string {
  if (seconds < 90) return `${seconds.toFixed(0)} s`;
  if (seconds < 5400) return `${(seconds / 60).toFixed(1)} min`;
  return `${(seconds / 3600).toFixed(1)} h`;
}
